File: 13. Roman to Integer/13. Roman to Integer.go
package main

import "fmt"

func romanToInt(s string) int {
	romanDict := map[string]int{
		"I": 1,
		"V": 5,
		"X": 10,
		"L": 50,
		"C": 100,
		"D": 500,
		"M": 1000,
	}
	var result int = 0
	for i, _ := range s {
		if i == len([]rune(s))-1 {
			break
		}
		if romanDict[string(s[i])] < romanDict[string(s[i+1])] {
			result -= romanDict[string(s[i])]
		} else {
			result += romanDict[string(s[i])]
		}
	}
	result += romanDict[string(s[len(s)-1])]
	return result
}

func main() {
	fmt.Println(romanToInt("III"))
	fmt.Println(romanToInt("LVIII"))
	fmt.Println(romanToInt("MCMXCIV"))
	fmt.Println(romanToInt("IX"))
}

File: 13. Roman to Integer/13. Roman to Integer.py
class Solution:
    roman_dict = {
        "I": 1,
        "IV": 4,
        "V": 5,
        "IX": 9,
        "X": 10,
        "XL": 40,
        "L": 50,
        "XC": 90,
        "C": 100,
        "CD": 400,
        "D": 500,
        "CM": 900,
        "M": 1000
    }
    def romanToInt(self, s: str) -> int:
        result = 0
        i = 0
        while i < len(s):
            if s[i] in ("I","X","C"):
                try:
                    result += self.roman_dict[s[i:i+2]]
                    i += 1
                except:
                    result += self.roman_dict[s[i]]
            else:
                result +=  self.roman_dict[s[i]]
            i += 1
        return result


if __name__ == '__main__':
    c = Solution()
    print(c.romanToInt("III"))
    print(c.romanToInt("LVIII"))
    print(c.romanToInt("MCMXCIV"))
    print(c.romanToInt("IX"))
